File: src/ai/internal/L1Model.go
package internal

import (
	board2 "github.com/HuXin0817/dots-and-boxes/src/board"
	"github.com/HuXin0817/dots-and-boxes/src/config"
	"github.com/HuXin0817/dots-and-boxes/src/model"
	"github.com/stretchr/testify/assert"
)

type L1Model struct {
	L0       L0Model
	auxBoard board2.BoardV3
}

func NewL1Model() *L1Model {
	return &L1Model{
		L0:       *NewL0Model(),
		auxBoard: *board2.NewBoardV3(),
	}
}

func (m *L1Model) BestCandidateEdges(b *board2.BoardV2) []model.Edge {
	if l := m.L0.BestCandidateEdges(b); !m.L0.EnemyUnscoreableEdges.Empty() || !m.L0.ScoreableEdge.Empty() {
		return l
	}
	mins := int(model.MaxBox + 1)
	Candidate := &m.L0.EnemyUnscoreableEdges
	if config.DEBUG {
		assert.True(nil, Candidate.Empty())
	}
	for _, e := range b.EmptyEdges() {
		m.auxBoard.Reset(&b.BoardV1)
		if config.DEBUG {
			assert.Equal(nil, m.auxBoard.Add(e), 0)
		} else {
			m.auxBoard.Add(e)
		}
		s := m.auxBoard.MaxObtainableScore(mins)
		if s < mins {
			mins = s
			Candidate.Reset(e)
		} else if s == mins {
			Candidate.Append(e)
		}
	}
	return Candidate.Export()
}

File: src/ai/L1Model.cpp
#include "L1Model.h"

std::span<const Edge>
L1Model::BestCandidateEdges(const BoardV2& board) {
  if (auto l = SubModel.BestCandidateEdges(board);
      !SubModel.EnemyUnscoreableEdges.Empty() || !SubModel.ScoreableEdges.Empty()) {
    return l;
  }

  int mins = Box::Max + 1;
  auto& Candidate = SubModel.EnemyUnscoreableEdges;
  assert(Candidate.Empty());

  for (auto e : board.EmptyEdges()) {
    AuxBoard.Reset(board.GetBoardV1());
    int v = AuxBoard.Add(e);
    assert(v == 0);

    int s = AuxBoard.MaxObtainableScore(mins);
    if (s < mins) {
      mins = s;
      Candidate.Reset(e);
    } else if (s == mins) {
      Candidate.Append(e);
    }
  }

  return Candidate.Export();
}
